## Quality Summary

```{r}
# Get corr_median
my_sample <- rownames(corr_data$annoNow)[which(corr_data$annoNow$Source != "RMapDB")]
corr_median <- data.frame("clean_name" = rownames(corr_data$annoNow), "R" = corr_data$corMat[my_sample,]) %>% 
    left_join(y = rmap_samples, by = "clean_name") %>%
    dplyr::filter(mode_group == "DRIP" & 
                    ! Condition %in% c("IgG", "RNaseH1", "RNASEH1", "RNH-high", "WKKD")) %>%
    pull(R) %>% median()

# Get annotations
annos <- anno_data %>%
    pivot_longer(!Annotation) %>%
    mutate(feature = paste0(Annotation, "__", name)) %>%
    dplyr::select(feature, value) 
annos <- as.data.frame(t(annos))
colnames(annos) <- as.character(unlist(annos[1,,drop=TRUE]))
annos <- annos[-1,]

# Pct aligned  
total_reads <- bam_stats$total_reads 
reads_aligned <- bam_stats$reads_aligned
pct_aligned <- 100*(reads_aligned/total_reads)
  
# rlfs
pval <- rlfs_data[[1]]$`regioneR::numOverlaps`$pval
pval <- -log10(pval)
minpvalpossible <- -log10(1/(1+rlfs_data[[1]]$`regioneR::numOverlaps`$ntimes))
zscore_summit <- rlfs_data[[1]]$`regioneR::numOverlaps`$zscore
if (is.na(zscore_summit)) {
  next
}
zscores <- rlfs_data[[2]]$`regioneR::numOverlaps`$shifted.z.scores
zscores <- (zscores - min(zscores)) + .01
zscore_freq <- round(zscores, 3) * 1000  # 1E5 precision
shifts <- rlfs_data[[2]]$`regioneR::numOverlaps`$shifts
distres <- unlist(lapply(seq(shifts), function(i) {rep(shifts[i], zscore_freq[i])}))
resrlfs <- as.data.frame(t(as.data.frame(unlist(fitdistrplus::descdist(distres, graph = FALSE)))))
colnames(resrlfs) <- paste0("rlfs_", colnames(resrlfs))
resrlfs$rlfs_pval <- pval

# Compile into mat
test_x <- annos
test_x$corr_median <- corr_median
test_x$pct_aligned <- pct_aligned
test_x$rlfs_kurtosis <- resrlfs$rlfs_kurtosis
test_x$rlfs_skewness <- resrlfs$rlfs_skewness
test_x <- as.matrix(test_x)
test_x <- apply(test_x, 1:2, as.numeric)

# XGBoost
xgb_feats <- c("corr_median", "Exon__Log2 Ratio (obs/exp)", "TTS__Log2 Ratio (obs/exp)", 
               "5UTR__Log2 Ratio (obs/exp)", "3UTR__Log2 Ratio (obs/exp)",
               "rlfs_kurtosis", "rlfs_skewness")


test_x_xgb <- test_x[,xgb_feats, drop = FALSE]
bst <- xgb.load(modelfile = xgb_file)
xgbprob <- predict(bst, newdata = test_x_xgb)
xgbverdict <- ifelse(xgbprob > .8, 'pass', 'fail')
xgbcolor <- ifelse(xgbprob > .90, 'green', ifelse(xgbprob > .70, 'orange', 'red'))
verdictcolor <- ifelse(xgbverdict == 'pass', 'green', 'red')

# Linear Regression
test_x <- as.data.frame(test_x)
colnames(test_x) <- gsub(colnames(test_x), pattern = " |\\(|\\)|\\/", replacement = "_")
colnames(test_x) <- gsub(colnames(test_x), pattern = "^([0-9]+)", replacement = "d\\1")
suppressWarnings(load(lm_file))
lmscore <- predict(lmfit, newdata = test_x) * 100
# lmscore <- ifelse(lmscore > 100, 100,
#                   ifelse(lmscore < 0, 0, lmscore))
lmcolor <- ifelse(lmscore > 90, 'green', 
                  ifelse(lmscore > 60, 'orange', 'red'))
```


<h4>Verdict: <strong style="color: `r verdictcolor`">`r toupper(xgbverdict)`</strong></h4>

<h4>Quality score: <strong style="color: `r lmcolor`">`r round(lmscore, 2)`</strong></h4>


### Comparison to RMapDB samples

Comparison of user-supplied sample to RMapDB DRIP-family samples (i.e., qDRIP, sDRIP, DRIPc, DRIP) based on quality score calculation detailed above.

```{r fig.height=5, fig.width=9}
rmap_quality_df <- suppressMessages(read_csv(rmap_quality)) 
rmap_quality_df$Source <- "RMapDB"
current_scores <- data.frame(sample_name, as.numeric(xgbprob > .5), xgbprob, as.numeric(xgbprob > .5), lmscore,
                             Source = "User-supplied")
colnames(current_scores) <- colnames(rmap_quality_df)
full_quality <- bind_rows(rmap_quality_df, current_scores)
full_quality <- full_quality[order(full_quality$lmscore),]
full_quality$rank <- seq(full_quality$lmscore)
full_quality$Verdict <- ifelse(full_quality$Source == "User-supplied", paste0("User-supplied: ", toupper(xgbverdict)), 
                               ifelse(full_quality$xgbverdict == 1, "PASS", "FAIL"))
# full_quality$Verdict <- ifelse(full_quality$xgbverdict == 1, "PASS", "FAIL")
full_quality$Verdict <- factor(full_quality$Verdict, levels = rev(c("PASS", "FAIL", paste0("User-supplied: ", toupper(xgbverdict)))))
full_quality <- full_quality[order(full_quality$Verdict, decreasing = TRUE),]
plt <- ggplot(data = full_quality, aes(x = rank, y = lmscore, 
                                text = clean_name, 
                                size = Source, alpha = Source,
                                color = Verdict)) +
  geom_point() +
  theme_bw(base_size = 15) +
  xlab("Rank") + ylab("Quality Score") + 
  labs(title = "Quality score comparison with RMapDB") +
  scale_color_manual(values = c("darkviolet", 'firebrick', 'forestgreen')) +
  scale_alpha_manual(values = c(.4, 1), guide=FALSE) +
  scale_size_manual(values = c(2, 4), guide=FALSE)
pltg <- ggplotly(plt, tooltip="text")
# From https://stackoverflow.com/questions/49133395/strange-formatting-of-legend-in-ggplotly-in-r
for (i in seq(length(pltg$x$data))){
  if (!is.null(pltg$x$data[[i]]$name)){
    pltg$x$data[[i]]$name =  gsub("\\(","",str_split(pltg$x$data[[i]]$name,",")[[1]][1])
  }
}
pltg
```


### Method details

This sample, `r sample_name`, received the verdict <strong style="color: `r verdictcolor`">`r toupper(xgbverdict)`</strong> from the XGBoost model which predicted that the probability of PASS for this sample was <strong style="color: `r xgbcolor`">`r round(100*xgbprob, 2)`%</strong>.

The linear regression model predicted the quality score of this sample to be <strong style="color: `r lmcolor`">`r round(lmscore, 2)`</strong>. 

The linear model follows this formula:

$y = 97.27C_m + 14.24I_r + 8.903S_r - 13.53L_r + 0.440A_p - 20.39$

Where:

$C_m = \text{Median pearson correlation among similar sample types}$

$I_r = \text{Intron enrichment in peaks log2(Observed / Expected)}$

$S_r = \text{SINE enrichment in peaks log2(Observed / Expected)}$

$L_r = \text{LINE enrichment in peaks log2(Observed / Expected)}$

$A_p = \text{percent of reads aligned to genome}$


\**See RSeq publication for additional details*

***
