#' @rdname RSeq_variables
#' @export
setClass("RSeq_variables",
         slots = list(
           sample_name = "character",
           out_name = "character",
           out_dir = "character",
           peak_available = "logical",
           bam_available = "logical",
           bam_file = "character",
           file_format = "character",
           fastq_available = "logical",
           fastq_1_available = "logical",
           fastq_2_available = "logical",
           fastq_dir = "character",
           mode = "character",
           fastq_QC_dir = "character",
           fastq_file = "character",
           fastq_file_1 = "character",
           fastq_file_2 = "character",
           genome_home_dir = "character",
           genome = "character",
           genome_fasta_file = "character",
           genome_annotations_file = "character",
           genome_fasta_file_exists = "logical",
           genome_annotations_file_exists = "logical",
           genome_fasta_url = "character",
           available_genome_indices = "character",
           genome_annotations_available = "logical",
           genome_annotations_url = "character",
           genome_available = "logical",
           cores = "integer",
           bam_dir = "character",
           bam_QC_dir = "character",
           peaks_dir = "character",
           full_genome_length = "numeric",
           effective_genome_size = "numeric",
           experiments = "character",
           controls = "character",
           moeity = "character",
           ip_type = "character",
           coverage_dir = "character",
           RSeq_version = "character",
           all_experiment_bams = "character",
           downstream_dir = "character",
           report_dir = "character",
           read_length = "numeric",
           paired_end = "logical",
           strand_specific = "logical",
           all_experiment_labels = "character"
         ))

